golang/cosmos/x/vtransfer/keeper/keeper.go

```
@@ -113,25 +214,29 @@ func (k Keeper) InterceptOnRecvPacket(ctx sdk.Context, ibcModule porttypes.IBCMo
 	if err != nil {
 		return channeltypes.NewErrorAcknowledgement(err)
 	}
-	ack := ibcModule.OnRecvPacket(ctx, strippedPacket, relayer)
 
+	portID := packet.GetDestPort()
+	channelID := packet.GetDestChannel()
+	if StorePacketData && !bytes.Equal(strippedPacket.GetData(), packet.GetData()) {
+		packetStore, packetKey := k.PacketStore(ctx, types.PacketDst, portID, channelID, packet.GetSequence())
+		packetStore.Set(packetKey, packet.GetData())
+	}
+
+	ack := ibcModule.OnRecvPacket(ctx, strippedPacket, relayer)
 	if ack == nil {
 		// Already declared to be an async ack.
 		return nil
 	}
-	portID := packet.GetDestPort()
-	channelID := packet.GetDestChannel()
 	capName := host.ChannelCapabilityPath(portID, channelID)
 	chanCap, ok := k.vibcKeeper.GetCapability(ctx, capName)
 	if !ok {
 		err := sdkerrors.Wrapf(channeltypes.ErrChannelCapabilityNotFound, "could not retrieve channel capability at: %s", capName)
 		return channeltypes.NewErrorAcknowledgement(err)
```

Are we not leaking the original packet data in our store in this case since the IBC msg server would be directly writing out this error ack result?

---

Correct, this path would bypass `vtransfer`, unlike our `ics4Wrapper` used by `ibcModule`.  So, I've changed `InterceptOnRecvPacket` to fail fast, before we store any data.  That prevents the leak.

---

From what I gather failing fast also causes the wrapped `ibcModule`'s `OnRecvPacket` to no longer be invoked. Is that expected? 

Thinking about it, is there a case where the `vibcKeeper` wouldn't have access to the destination channel if the `vtransfer` module does? If that can't happen, I like the new order better as it makes it more clear where hooking happens.

---

All the IBC capabilities used by `vtransfer` are stored in `k.vibcKeeper`, so `vtransfer` can't have more access than `vibc` does.